Psalm support (+general PHPDoc fixes)

alexeyshockov opened the pull request:
- [x] Psalm config
- [x] Psalm baseline file (if there are still errors after my quick fixes)
- [x] `number` → `int|float` ([PHPDoc standard doesn't contain such a type](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#appendix-a-types))
- [x] Psalm's `@param-out`
- [ ] `@psalm-immutable` (for classes) and `@psalm-pure` (for static factory methods) — probably in a separate PR

alexeyshockov commented:
@BenMorel, most of the fixes make sense for all static analyzers (Psalm, PHPStan, Phan).

What do you think?

BenMorel commented on src/Internal/Calculator/BcMathCalculator.php:
```diff
@@ -42,15 +43,25 @@ public function mul(string $a, string $b) : string
+        $result = \bcdiv($a, $b, 0);
+        if ($result === null) {
+            throw DivisionByZeroException::divisionByZero();
+        }
```
These assertions are not necessary. The Calculator interface [explicitly states](https://github.com/brick/math/blob/15ebb638d309aff44a016f3fcf8c2d83b7bef611/src/Internal/Calculator.php) that all inputs must be validated, or the result is undefined. The applies to `Calculator::divQ()`, which states:

> The divisor, must not be zero.

This is for performance, and is indirectly largely covered by the large number of tests!

alexeyshockov replied:
Thanks, didn't notice that. Removed the check, put the error is the baseline file for now (we can cast the function result to string to get rid of the error, but I don't want to do it just for static analyzer).

BenMorel commented on src/BigDecimal.php:
```diff
@@ -770,10 +770,12 @@ public function unserialize($value) : void
-     * @param string     $b A variable to store the scaled integer value of $y.
+     * @param BigDecimal  $x The first decimal number.
+     * @param BigDecimal  $y The second decimal number.
+     * @param string|null $a A variable to store the scaled integer value of $x.
```
Why is this `string|null`? I understand that when the method is called without `$a` being set, it's `null` as a result, but I really feel like it makes the documentation misleading. Is `@param-out` acceptable without `@param` at all?

alexeyshockov replied:
Reverted `null` changes. With `@param-out` is works just fine.

BenMorel commented:
Hi 👋

> Psalm config

I'm quite happy to introduce a static analyzer to this project (it should also be added to the Travis CI build process in this case), however I'm interested to know what lead you to choose Psalm, over phpstan, that seems to be used by quite a number of OSS projects?

> number → int|float (PHPDoc standard doesn't contain such a type)

You seem to be right. PHPStorm happily accepts it, and it made each method description a bit less verbose, so I chose to use it. I should probably revisit this, if PHPStorm is the only one to accept this, and it's not part of any other standard.

> Psalm's @param-out

Unless there's a very good reason to choose Psalm over any other static analyzer, and I can be confident that it's here to stay, I'm a bit reluctant to start adding Psalm-specific annotations in the code!

alexeyshockov commented:
Thanks for your feedback.

> I'm interested to know what lead you to choose Psalm, over phpstan, that seems to be used by quite a number of OSS projects?

It's not a hard preference, I've just found Plalm a go to solution for my projects, because it has more powerful type system (ability to describe closures, support for templates aka generics,..). It's super beneficial for some (see [lstrojny/functional-php](https://github.com/lstrojny/functional-php/tree/psalm)) and not a big deal for others (this library, for example, doesn't use closures or generic containers a lot). My overall impression is that Psalm is more feature rich, allows you to add additional meta information to the code to do more complex static checks on it.

For the support I don't worry: the project is there for years already and evolving well, along with PHPStan and Phan, two other main static analyzers (thanks to @muglug for his great work).

Anyway, for this kind of job you are not limited to only one tool. Psalm and PHPStan can be used together to ensure even better code quality, if you want so.

> Unless there's a very good reason to choose Psalm over any other static analyzer, and I can be confident that it's here to stay, I'm a bit reluctant to start adding Psalm-specific annotations in the code!

Especially in this place `@param-out` tag really shines, I don't know another way to save all the type checks and get rid of the related errors (PHPStan simply doesn't find this issue). IMO it's worth nothing to include such small thing in the code to pass the type checks.

I also fixed the comments, and added static analysis step to the CI process. Hope you we are thinking in the same direction :)

muglug commented:
> and I can be confident that it's here to stay

It's not going anywhere – I have no plans to stop work on it anytime soon and, if I were to fall under a bus, there are plenty of developers who can carry on the work (both at Vimeo and elsewhere).

muglug commented on src/Internal/Calculator.php:
```diff
@@ -93,9 +93,13 @@ private static function detect() : Calculator
      * @param string $a    The first operand.
      * @param string $b    The second operand.
      * @param string $aDig A variable to store the digits of the first operand.
+     * @param-out string
```
These `@param-out` annotations are unnecessary for Psalm (and also invalid as they lack a param name – Psalm emits `InvalidDocblock` issues for that).

Psalm assumes that any by-reference params retain their type when passed into a function. If that function _actually_ changes the type (e.g. here: https://psalm.dev/r/983088573a) then you need a `@param-out` annotation. But here, since the type doesn’t change, it's fine.

alexeyshockov replied:
Thanks! I don't know why, but I don't get `InvalidDocblock` locally...

Anyway I changed the code to properly use `@param-out`. The out type is different from the input here, because in the input `null` is allowed.

An example of the error: https://psalm.dev/r/07a893b434
Properly annotated: https://psalm.dev/r/59893b4e3e

@BenMorel, I understand that `string|null` might be a bit confusing, but it's the correct type annotation here. I can do
```php
/**
 * @param string $s
 * @psalm-param string|null $s
 * @param-out string $s
 */
```
but I really don't want to.

BenMorel replied:
Actually you can use `@param null` here: https://psalm.dev/r/6d0dfb8a57
Together with `@param-out string $s`

alexeyshockov replied:
True. Changed the PR to `@param null`, please review :)

BenMorel commented on .travis.yml:
```diff
@@ -17,6 +17,7 @@ before_script:
 script:
   - mkdir -p build/logs
   - vendor/bin/phpunit --coverage-clover build/logs/clover.xml
+  - vendor/bin/psalm --show-info=false --no-progress
```
Right now, this is run after every single build in the matrix, which is a waste of resources. There should be a job dedicated to running the static analyzer, and no PHPUnit tests. See for example: https://travis-ci.org/doctrine/dbal

alexeyshockov replied:
Good idea. Done. I also moved code coverage report to this "only once" section.

BenMorel replied:
Careful: coverage has to be performed on every job, as different calculators will be covered by different tests!

Also, please don't use a "Smoke testing" stage, this can belong to the main test stage. Doctrine uses smoke testing to avoid running the main test suite if the smoke test fails, but this is not what I had in mind here.

BenMorel commented:
@muglug 

> It's not going anywhere – I have no plans to stop work on it anytime soon and, if I were to fall under a bus, there are plenty of developers who can carry on the work (both at Vimeo and elsewhere).

I'm not really worried about the future of Psalm itself, but more about whether or not Psalm is the static analyzer that I'm going to use, and keep, on this project! Though after reading a few comparisons, it does indeed seem like a good choice. And please don't fall under a bus! ;)

@alexeyshockov 

I'm happy to try to introduce a static analyzer, as I definitely think it can help with writing better code, and this is something I think makes sense to use in the CI process together with unit tests.

So far however, on this codebase, I'm not that impressed with what it's been able to uncover:

- change `number` to `int|string` in PHPDoc
- somehow better document (subjective: with psalm-specific annotations) the output type of by-reference parameters
- ...nothing else so far?

And there are many ignored false positives in `psalm-baseline.xml`; ideally, if I'm introducing a static analyzer, I wouldn't like to ignore errors at all; furthermore, not in an external config file (worst case scenario, I *could* accept a few inline annotations here and there to hint the static analyzer that this is done on purpose).

There are actually quite a few examples where the static analyzer struggles:

- higher-level validation assumptions in calculators: for example [bcdiv() can return null, but will never do so](https://github.com/brick/math/blob/15ebb638d309aff44a016f3fcf8c2d83b7bef611/src/Internal/Calculator/BcMathCalculator.php#L45) because the input arguments to calculators are pre-validated.
- [addition of numeric strings](https://github.com/brick/math/blob/15ebb638d309aff44a016f3fcf8c2d83b7bef611/src/Internal/Calculator/NativeCalculator.php#L357): as weird as it looks, [this has actually been done on purpose](https://github.com/brick/math/commit/0aeb6c0bc7e1f06dde3857aa85b505b9cbdb6080). We know for a fact that operands are numeric strings, and that PHP will implicitly convert them. We therefore remove extra opcodes for explicit casting prior to addition, and (very slightly) improve performance.
- [division guaranteed to yield an integer](https://github.com/brick/math/blame/master/src/Internal/Calculator/NativeCalculator.php#L222): from the lines above, it's clear that the division will never yield a `float`, but Psalm cannot understand this
- [Variable $i has already been assigned in a for/foreach loop](https://github.com/brick/math/blob/master/src/Internal/Calculator/NativeCalculator.php#L415): Psalm just didn't understand the intent of the loop.

All in all, even though I was quite excited to try a static analyzer, as it stands, I can't help but think that on this project, the clutter outweighs the benefits?

alexeyshockov commented:
> ideally, if I'm introducing a static analyzer, I wouldn't like to ignore errors at all; furthermore, not in an external config file (worst case scenario, I could accept a few inline annotations here and there to hint the static analyzer that this is done on purpose).

I agree, but my point is to move step by step. I did obvious changes/annotations and put the remaining (small) things in below the besaline. The next PR will be about getting rid of the baseline.

IMO it's better than doing all the things in one huge PR. Users (me!) will benefit a lot already. Really, I started all this work initially just because of the `number` type inconsistency, because I'm getting errors from Psalm in my app, using the current stable version of the lib.

> All in all, even though I was quite excited to try a static analyzer, as it stands, I can't help but think that on this project, the clutter outweighs the benefits?

Low error amount in the current codebase is not a sign of "dumb analyzer", but the expected result for a **relatively small**, **mature** project that **hasn't been changing much in the last months**. Basically, if there are errors, you should have already gotten issue reports/PR with fixes, because there are a lot of users of the package.

@BenMorel, IMO the main reason to integrate a static analyzer is to proactively prevent bugs in the future. (Also pay attention to my plans to integrate Psalm's `@psalm-pure` and `@psalm-immutable` annotations, it will make immutability statement from the README a _checked_ guarantee, not just a text sentence).

BenMorel commented:
Thanks for your comment. Moving forward, how can we "fix" the errors that are currently muted in the baseline? You're right that this can belong to another PR, but I'd like to get a preview of the big picture, to understand which kind of modifications the codebase would have to go through!

muglug commented:
> Variable $i has already been assigned in a `for/foreach` loop: Psalm just didn't understand the intent of the loop.

I’m going to kill that check entirely — it felt smart at the time, but I’ve only ever seen it find a couple of bugs, with tons of false-positives.

muglug commented:
I agree wholeheartedly with @alexeyshockov about the use of static analysis in a small, well-tested project.

PHPUnit has used Psalm for a year or so, but not because it has many bugs or Sebastian Bergmann is a bug-prone developer. Rather, it helps increase the quality of submitted PRs, and can be used as a useful code-quality metric (via a type coverage button hosted at https://shepherd.dev)

BenMorel commented:
> I’m going to kill that check entirely — it felt smart at the time, but I’ve only ever seen it find a couple of bugs, with tons of false-positives.

I'm not sure what the scope of the check is, but PHPStorm has a similar inspection that only triggers if you attempt to use the same variable twice in nested loops:

```php
for ($i = 0; $i < 10; $i++) {
    for ($i = 0; $i < 100; $i++) {
        ...
    }
}
```

And this one is occasionally useful, and does not trigger on the particular case above.

BenMorel commented:
@alexeyshockov I'd like to move forward with this PR. Could you please investigate the following points:

- Smoke testing should run [before](https://travis-ci.org/doctrine/dbal/builds/639730541) tests, not [after](https://travis-ci.org/brick/math/builds/637398068). It looks like putting them in the right order is the only way to prevent another stage from starting if a previous stage failed.
- Test coverage [decreased](https://coveralls.io/builds/28123097) from nearly 100% to 88%. I'm surprised as you've correctly set `COVERAGE=yes` for each of the 3 calculators on PHP 7.4, so I have no explanation to offer at the moment.

Bonus question:

I like your CI speedup by disabling coverage + xdebug for most tests! But **should we run coverage in smoke testing?** Or should we speedup the smoke testing instead, and run coverage afterwards?

This is just a question, I don't have a strong opinion here.

Pros:

- faster smoke testing

Cons:

- no coverage info displayed in PRs if smoke testing fails (but does it matter then?)
- probably more complicated Travis config; this would probably require to further split the test stage:
    - 1st stage: smoke testing on a single PHP version with no coverage
    - 2nd stage: tests with:
        - coverage enabled on a single PHP version
        - coverage disabled on remaining PHP versions

What's your opinion here?

alexeyshockov commented:
Sorry for the delay here.

> Smoke testing should run before tests, not after. It looks like putting them in the right order is the only way to prevent another stage from starting if a previous stage failed.

Done

> Test coverage decreased from nearly 100% to 88%. I'm surprised as you've correctly set COVERAGE=yes for each of the 3 calculators on PHP 7.4, so I have no explanation to offer at the moment.

Yes, that's also what I noticed and cannot understand.

> I like your CI speedup by disabling coverage + xdebug for most tests! But should we run coverage in smoke testing? Or should we speedup the smoke testing instead, and run coverage afterwards?

I would say that maybe it's better to move code coverage generation away from the smoke testing stage. So we will have:
- smoke tests (all calculators on the latest PHP version (7.4 here), without coverage)
- tests (all calculators on all other PHP versions, with coverage enabled only for ONE PHP version)

Smoke testing aims to be fast, to ensure that we can go further... Then all the usual combinations. And we can enable the coverage report only on the lowest PHP version, for example. It wouldn't complicate Travis config a lot.

BenMorel commented on .travis.yml:
```diff
@@ -1,22 +1,62 @@
 
+stages:
+  - Smoke Testing
+  - Test
```
Shouldn't this include "Test (with coverage)" as well?

alexeyshockov replied:
All stages that are not covered here runs after, implicitly. But I added all the stages explicitly, so done.

BenMorel commented:
All good. Thanks! 👍 

> Yes, that's also what I noticed and cannot understand.

It's actually not related to your PR. I've sent an email to Coveralls to understand what's happening.
